apps/astarte_appengine_api/lib/astarte_appengine_api_web/router.ex

```diff
@@ -19,106 +19,172 @@ defmodule Astarte.AppEngine.APIWeb.Router do
-    get "/groups/:group_name/devices/:device_id", DeviceStatusByGroupController, :show
+      post "/devices/:device_id/interfaces/:interface/*path_tokens",
+           InterfaceValuesByGroupController,
+           :update
 
-    patch "/groups/:group_name/devices/:device_id", DeviceStatusByGroupController, :update
+      delete "/devices/:device_id/interfaces/:interface/*path_tokens",
+             InterfaceValuesByGroupController,
+             :delete
+    end
 
-    get "/groups/:group_name/devices/:device_id/interfaces",
-        InterfaceValuesByGroupController,
-        :index
+    scope "/devices-by-alias/:device_alias" do
+      pipe_through :interface_values_device_group_api
 
-    get "/groups/:group_name/devices/:device_id/interfaces/:interface",
-        InterfaceValuesByGroupController,
-        :show
+      resources "/interfaces",
+                InterfaceValuesByDeviceAliasController,
+                only: [:index, :show],
+                param: "interface"
 
-    get "/groups/:group_name/devices/:device_id/interfaces/:interface/*path_tokens",
-        InterfaceValuesByGroupController,
-        :show
+      get "/interfaces/:interface/*path_tokens",
+          InterfaceValuesByDeviceAliasController,
+          :show
 
-    put "/groups/:group_name/devices/:device_id/interfaces/:interface/*path_tokens",
-        InterfaceValuesByGroupController,
-        :update
+      put "/interfaces/:interface/*path_tokens",
+          InterfaceValuesByDeviceAliasController,
+          :update
 
-    post "/groups/:group_name/devices/:device_id/interfaces/:interface/*path_tokens",
-         InterfaceValuesByGroupController,
-         :update
+      post "/interfaces/:interface/*path_tokens",
+           InterfaceValuesByDeviceAliasController,
+           :update
 
-    delete "/groups/:group_name/devices/:device_id/interfaces/:interface/*path_tokens",
-           InterfaceValuesByGroupController,
-           :delete
+      delete "/interfaces/:interface/*path_tokens",
+             InterfaceValuesByDeviceAliasController,
+             :delete
+    end
+
+    scope "/devices-by-alias" do
+      pipe_through :device_alias_api
+
+      resources "/", DeviceStatusByAliasController,
+        only: [:index, :show, :update],
+        param: "device_alias"
+    end
+
+    scope "/groups" do
+      pipe_through :device_group_api
+
+      get "/", GroupsController, :index
+      post "/", GroupsController, :create
+      get "/:group_name", GroupsController, :show
+      post "/:group_name/devices", GroupsController, :add_device
+      delete "/:group_name/devices/:device_id", GroupsController, :remove_device
+    end
+
+    scope "/groups" do
+      pipe_through :device_group_name_api
+
+      get "/:group_name/devices", DeviceStatusByGroupController, :index
+      get "/:group_name/devices/:device_id", DeviceStatusByGroupController, :show
+
+      patch "/:group_name/devices/:device_id", DeviceStatusByGroupController, :update
+    end
```

Comment:
It feels strange to have the same scope declared multiple times.
I guess that you were trying to separate routes depending on the `pipe_through` pipeline.

However I would first start declaring the scopes, and nested scopes. Then I would add the appropriate plugs to each level of the nested scopes.
It is not necessary to use pipelines, we can also use plugs directly.

Also note that:
- both the pipelines `device_group_api` and `device_group_name_api` logs the group name. But we don't have a group name to log for GET /groups and POST /groups.
- only the pipeline `device_group_name_api` logs the device ID. But we also have a route with a device ID in the other scope, where the device ID is not logged.

I would suggest first declaring the structure of the routes thanks to nested `scope` blocks, and then adding the appropriate plugs where needed.
For example:

```elixir
scope "/groups" do
  get "/", GroupsController, :index
  post "/", GroupsController, :create

  scope "/:group_name" do
    plug Astarte.AppEngine.APIWeb.Plug.LogGroupName

    get "/", GroupsController, :show
    get "/devices", DeviceStatusByGroupController, :index
    post "/devices", GroupsController, :add_device

    scope "/devices/:device_id" do
      plug Astarte.AppEngine.APIWeb.Plug.LogDeviceId
      
      get "/", DeviceStatusByGroupController, :show
      patch "/", DeviceStatusByGroupController, :update
      delete "/", GroupsController, :remove_device
    end
  end
end
```

Reply:
Thank you for the review and suggestions. But, I have a problem with implementing this suggestion. When run the command ` mix phx.router `  returns an error plug `(RuntimeError) cannot define plug at the router level, plug must be defined inside a pipeline`